sdk/metric/meter_test.go

```diff
@@ -998,37 +1007,282 @@ func TestAttributeFilter(t *testing.T) {
+		observations[thread1] = 53
+		observations[thread2] = 38
+
+		*wantFiltered = 91
+		*wantThread1 = 53
+		*wantThread2 = 38
+
+		verify(t)
+
+		// During the time range (T2, T3]
+		//     pid = 1001, tid = 1, #PF = 56
+		//     pid = 1001, tid = 2, #PF = 42
+		observations[thread1] = 56
+		observations[thread2] = 42
+
+		*wantFiltered = 98
+		*wantThread1 = 56
+		*wantThread2 = 42
+
+		verify(t)
+
+		// During the time range (T3, T4]:
+		//     pid = 1001, tid = 1, #PF = 60
+		//     pid = 1001, tid = 2, #PF = 47
+		observations[thread1] = 60
+		observations[thread2] = 47
+
+		*wantFiltered = 107
+		*wantThread1 = 60
+		*wantThread2 = 47
+
+		verify(t)
+
+		// During the time range (T4, T5]:
+		//     thread 1 died, thread 3 started
+		//     pid = 1001, tid = 2, #PF = 53
+		//     pid = 1001, tid = 3, #PF = 5
+		delete(observations, thread1)
+		observations[thread2] = 53
+		observations[thread3] = 5
+
+		*wantFiltered = 58
+		want.Metrics[1].Data = metricdata.Sum[int64]{
+			Temporality: temporality,
+			IsMonotonic: true,
+			DataPoints: []metricdata.DataPoint[int64]{
+				// Thread 1 remains at last measured value.
+				{Attributes: thread1, Value: 60},
+				{Attributes: thread2, Value: 53},
+				{Attributes: thread3, Value: 5},
+			},
+		}
+
+		verify(t)
+	})
+
+	t.Run("Delta", func(t *testing.T) {
+		temporality := metricdata.DeltaTemporality
+		observations, verify, want, wantFiltered, wantThread1, wantThread2 := setup(t, temporality)
+
+		// During the time range (T0, T1]:
+		//     pid = 1001, tid = 1, #PF = 50
+		//     pid = 1001, tid = 2, #PF = 30
+		observations[thread1] = 50
+		observations[thread2] = 30
+
+		*wantFiltered = 80
+		*wantThread1 = 50
+		*wantThread2 = 30
+
+		verify(t)
+
+		// During the time range (T1, T2]:
+		//     pid = 1001, tid = 1, #PF = 53
+		//     pid = 1001, tid = 2, #PF = 38
+		observations[thread1] = 53
+		observations[thread2] = 38
+
+		*wantFiltered = 11
+		*wantThread1 = 3
+		*wantThread2 = 8
+
+		verify(t)
+
+		// During the time range (T2, T3]
+		//     pid = 1001, tid = 1, #PF = 56
+		//     pid = 1001, tid = 2, #PF = 42
+		observations[thread1] = 56
+		observations[thread2] = 42
+
+		*wantFiltered = 7
+		*wantThread1 = 3
+		*wantThread2 = 4
+
+		verify(t)
+
+		// During the time range (T3, T4]:
+		//     pid = 1001, tid = 1, #PF = 60
+		//     pid = 1001, tid = 2, #PF = 47
+		observations[thread1] = 60
+		observations[thread2] = 47
+
+		*wantFiltered = 9
+		*wantThread1 = 4
+		*wantThread2 = 5
+
+		verify(t)
+
+		// During the time range (T4, T5]:
+		//     thread 1 died, thread 3 started
+		//     pid = 1001, tid = 2, #PF = 53
+		//     pid = 1001, tid = 3, #PF = 5
+		delete(observations, thread1)
+		observations[thread2] = 53
+		observations[thread3] = 5
+
+		*wantFiltered = -49
+		want.Metrics[1].Data = metricdata.Sum[int64]{
+			Temporality: temporality,
+			IsMonotonic: true,
+			DataPoints: []metricdata.DataPoint[int64]{
+				// Thread 1 remains at last measured value.
```

Comment:
Is this correct?  Seems to differ from the example in the supplementary guidelines: https://github.com/open-telemetry/opentelemetry-specification/blob/25f75bfde209630c60728039f27c041f70e5e738/specification/metrics/supplementary-guidelines.md#asynchronous-example-cumulative-temporality.  I'm trying to figure out where the spec says this

Reply:
I take the omission in the supplementary guidelines as not specifying what happens to the exported data for thread 1. For our SDK this goes back to the question of how to handle resets or dropped attribute sets, https://github.com/open-telemetry/opentelemetry-go/issues/3006. Until #3006 and the related https://github.com/open-telemetry/opentelemetry-specification/issues/1891#issuecomment-1194348335 are resolved, we don't have a way to persist state across collection cycles but also allow resets. And the delta implementation requires maintaining state across collections cycles to calculate the cumulative-to-delta conversion.

Reply:
> And the delta implementation requires maintaining state across collections cycles to calculate the cumulative-to-delta conversion.

Does this imply the cumulative implementation also needs to maintain state in this case?  Haven't looked to see if this is possible implementation-wise.

Reply:
> Does this imply the cumulative implementation also needs to maintain state in this case? Haven't looked to see if this is possible implementation-wise.

It does, but only to ensure to false reset gaps are introduced.

I guess, given how the measured values are precomputed and set, you could expect the previous cycle's value to be reported again and wouldn't need to maintain state. But if we relied on this and the user failed to report the value for a cycle or two and then continued reporting it would introduce gaps (or zeros) in the reported cumulative metric value.

Reply:
> I guess, given how the measured values are precomputed and set, you could expect the previous cycle's value to be reported again and wouldn't need to maintain state. But if we relied on this and the user failed to report the value for a cycle or two and then continued reporting it would introduce gaps (or zeros) in the reported cumulative metric value.

For what its worth, this would also diverge from our other non-pre-computed cumulative sum and mean we had inconsistent output behavior for a `Counter` versus an `ObservableCounter`.

Reply:
I think there's a potential for specification work on this question, but here's how I've been thinking about it. I feel that the reason for asynchronous instruments is to give the application a performance benefit when reporting metrics. Therefore, we should promote that benefit by not requiring the SDK to maintain additional state to ensure some kind of correctness and order.

It is the user's responsibility to behave, I think, which means asynchronous callbacks should not "forget" or duplicate attribute sets they report. The SDK has to go to some lengths to de-duplicate inputs and track forgotten attribute sets, and until convinced otherwise I think these should remain as unspecified behaviors.

Reply:
> For what its worth, this would also diverge from our other non-pre-computed cumulative sum and mean we had inconsistent output behavior for a Counter versus an ObservableCounter.

I think this depends on how you define consistent user behavior with a Counter vs ObservableCounter.  We could say that incrementing a Counter is equivalent to an ObservableCounter that keeps returning 1 value higher than prior.  If, someday, the synchronous counter had a Delete(labels) function, the equivalent behavior in ObservableCounter would be to stop returning observations with those labels.  Today, we don't have an equivalent in the synchronous case, which I think is OK (prom, for example, allows removing label sets in the async case, but not in the sync case).

> It is the user's responsibility to behave, I think, which means asynchronous callbacks should not "forget" or duplicate attribute sets they report. The SDK has to go to some lengths to de-duplicate inputs and track forgotten attribute sets, and until convinced otherwise I think these should remain as unspecified behaviors.

+1.  To spell out my desired behavior further, IMO an async counter which stops reporting a label set (ignore errors for now) should not be reported in the cumulative or delta case, and any memory/caches associated with that label set should be immediately released.  The next observation of that label set should be reported as if it were completely new.  The start time should be the time Collect was previously invoked, since that was the last time we knew this series was absent.

> it would introduce gaps (or zeros) in the reported cumulative metric value.

This feels correct to me.  It gives the user exactly what they've asked for, which is no values during a period.

Reply:
I don't have strong feelings here.

@dashpole can you track this in a new issue? It doesn't seem like something that should block this PR given it is about default behavior and the PR is addressing filters, right?

Reply:
This 100% shouldn't block this PR.  Sorry if that was not apparent

Reply:
Left https://github.com/open-telemetry/opentelemetry-specification/issues/1891#issuecomment-1396167105

Reply:
@dashpole I agree w/ your summary, thanks.

Reply:
Here is an issue on our repo as well: https://github.com/open-telemetry/opentelemetry-go/issues/3605